PyTDF/MapReduce.py

```diff
@@ -0,0 +1,135 @@
+class MapReduce(object):
+    """
+    Class for performing 
+    MapReduce on datasets
+
+    """
+    def __init__(self):
+        
+        self.actions = 0
+        self.tfs = 0
+        self.operations = []
+        self.action_node_map = {}
+
+    def execute_from_functions(self, mapper, reducer):
+        """
+        Use DistROOT to execute a distributed job
+        with the given mapper and reducer functions
+
+        TODO(shravan97)
+        """
+
+        pass
+
+    def _dfs(self, node, prev = 't', ops = []):
+        """
+        Method that does a depth-first
+        traversal and takes note of the 
+        necessary operations to carry out
+
+        """
+        
+        ops_copy = list(ops)
+
+        if not node.operation: # If root
+            for n in node.next_nodes:
+                self._dfs(n)
+            return
+
+        if node.operation.op_type == "a":
+            self.actions+=1
+            
+            ops_copy.append(node.operation)
+            
+            new_var_name = "ta"+str(self.actions)
+            self.operations.append((new_var_name, prev, ops_copy))
+            self.action_node_map[new_var_name] = node
+
+        elif node.operation.op_type == "t":
+            if len(node.next_nodes) > 1:
+                self.tfs+=1
+
+                ops_copy.append(node.operation)
+                
+                cur = 'tt'+ str(self.tfs)
+                self.operations.append((cur, prev, ops_copy))
+                new_tuple = (cur, [])
+            
+            elif len(node.next_nodes) == 1:
+                ops_copy.append(node.operation)
+                new_tuple = (prev, ops_copy)
+
+            else:
+                self.tfs += 1
+                
+                ops_copy.append(node.operation)
+                
+                cur = 'tt' + str(self.tfs)
+                self.operations.append((cur, prev, ops_copy))
+
+            for n in node.next_nodes:
+                self._dfs(n, *new_tuple)
+
+    def execute_from_graph(self, root_node):
+        """
+        Method that calls the _get_mapper and 
+        _get_reducer methods to pass the mapper
+        and reducer functions to DistROOT
+
+        TODO(shravan97)
+        """
+
+        # from DistROOT import DistTree
+        # dTree = DistTree(root_node.file_list, root_node.treename, root_node.npartitions)
+        # reduced_dict = dTree.ProcessAndMerge(self._get_mapper(root_node), self._get_reducer())
+        # for val in reduced_dict:
+        #     self.action_node_map[val].value = reduced_dict[val]
+        
+        pass
+
+    def _get_mapper(self, root_node):
```

Comment:
I would not place this function in a `MapReduce` module. This function should be a utility for you to get a function that, given a TDataFrame, runs the whole graph. I would not even call it `_get_mapper` but rather `get_callable` or something like that.

You will always call this function when triggered by the access to `Proxy` object and, depending on the backend you are using (local, distributed) you will get its result and do whatever you need to do. In the local backend, you will just construct the initial TDataFrame and call the function. In the distributed, you will call ProcessAndMerge. For the moment, I would forget about the latter and about merging and just make it work local. But it is important that we separate things in a way that makes sense conceptually.